### DISCUSSION TITLE
What causes the car to send an update?

### DISCUSSION THREAD

**Discussion by busywait** [Hyundai-Kia-Connect/kia_uvo#868] [Q&A]
I'm trying to understand how fresh the data provided by this integration can be when I configure it with Forced Refresh Interval = 9999, and what extra data I get when I use automation to call the integration "force refresh" service call.

I've noticed that sometimes my car updates its location even though the integration's "Last Updated At" sensor is unchanged. Does anyone know what needs to be happening in the car to send it's location (I'm in the UK with EU software in the car head unit).

I *think* that the locations get sent when I turn the car off. I've got an automation that calls the Force Refresh service when the location changes so that I keep the full set of car data (including battery levels) updated each time the car gets used.

I'd prefer it if the location was updated when the car gets turned on, then I could set up an automation to track each journey - but I guess it just doesn't 🤷‍♀️

Also, I have charging alerts setup (charge completed and 10 minutes before charge completed) in the car/Kia app. Do these cause new data to be sent from the car to the cloud servers? It seems a bit unpredictable whether this integration catches changes in battery level (maybe something to do with the blackout "No Force Refresh" settings?), but often it looks like the current battery level is sent at 10%, 20%, 30%,  etc

I don't have any automated refresh triggered by the battery level changing, but I do trigger a full refresh when my home chargepoint is unplugged from the car.

With those two automations, I get a reasonably up-to-date view of the car's state without having to do any scheduled polling that would keep the car electronics awake and draw power from the little 12V battery.

**Comment by brandmail** [comment]
I have a new car with kia connect, and I have the same issue. The data is far from live. The problem is that they have limit the data request by 200 in our country, to avoid battery problems with the car. Each data request take some enegry from the cars battery. So thats why they have limits on that.

**Comment by bobthebuilder4711** [comment]
The location is updated every time the car is turned off. I'm guessing this is so you can find it again via the app without having to poll the car.

Our last car was a Kia Cee'd Hybrid and that used to update every 30 minutes or so while charging. Or so I thought. Now reading about the updates getting sent at 10%, 20%, 30% and so on and that makes total sense. The small battery of the hybrid likely produces more frequent updates compared to a full blown EV with a much larger battery while charging at the same speed.

And yes, I can confirm with our new Kia Niro EV that the updates do in fact come in once the battery reaches a full 40%, 50%, etc. Haven't drained it much lower, but I'm guessing also at 10%, 20% etc.

Now I wonder what other sensors I should change to test what forces an update.

You don't really need to know the current SoC if the car has only be sitting around. So that's a plus. But you need to know how much the battery has been drained from the last trip. So a full update after turning off the car would indeed be nice. 

That way you would be able to see the last SoC before starting to charge. From that you could estimate the SoC based on the kWh counted by the charger. (I'm using a 1-phase/3-phase charging brick with up to 11kW, but it doesn't ask the car for the SoC, so that has to come in via Kia Connect.) And that way you would have an up-to-date value for the SoC even if it didn't come from the car directly.

Only problem is if the limit is really at 200 minutes and you take a trip that is shorter than 3 hours and can't force an update.

**Reply by busywait** [reply]
"if the limit is really at 200 minutes"
The limit is 200 (or whatever number) of requests to the Web Api each day - plenty for my automations.

I do an update from the Web api cache every 15 minutes ("scan interval" in the integration config). 

To avoid drain on the car 12V battery I minimise the number of forced updates from the integration when the car is off: I've got one automation that does a forced update every 12 minutes *while the API shows engine is on*, one forced update is triggered when my smart charger detects it has been connected, and I don't do any scheduled forced updates.

[edit: my automations also do a full forced update when the location changes, because location updates don't usually include the battery status]

I've found that fairly reliable except that
- the first forced update after the engine/car is turned off often returns a stale location
- a manually triggered event is usually needed to detect the engine/car is on, for example manually triggering a forced update or using the Kia Connect app status refresh

**Reply by bobthebuilder4711** [reply]
Oh, okay. Yes, 200 requests per day are plenty.

I'm fine with only getting a fresh update once the car is shut off. This is mainly for display in our kitchen/living room and of course for solar excess charging, so if the car is turned on and driving around, we don't need those two anyway.

Would you be so kind to show us the yaml of your automation that is triggered by the location update? Is it just this?:
```
triggers:
  - trigger: state
    entity_id:
      - device_tracker.niro_location
```

**Reply by busywait** [reply]
I was learning about HA automations and the Kia integration when I wrote the location change automation so I think it's too complicated, is also triggered by engine ON and OFF transitions, and relies on at least one timer, a YAML template sensor "car_gps", and a script that rate limits forced updates. It does work, but isn't a helpful example. The comment at the top might be helpful. 

By the time I got the template binary_sensor.car_gps to work to indicate the car was moving I realised that the simple state change that you've used was probably sufficient for anything practical, but I've never got around to simplifying.

`alias: Home EV Start and End Journey Triggers
description: >-
  When the car is turned off then a location update is sent. The updates come
  without any status information about the car. So, assume that changing
  locations, however small, indicate the car is recently active and a forced update will
  not cause excessive power draw. April 2024: stale location data is sent in the
  first forced update after the engine is turned off, and calling another forced
  update gets the correct location.
mode: single
triggers:
...
  - entity_id:
      - binary_sensor.car_gps
    from: "off"
    to: "on"
    id: GPSOn
    enabled: true
    trigger: state
conditions:
  - condition: state
    entity_id: timer.home_ev_usage_countdown
    state: idle
    enabled: true
actions:
  - choose:
      - conditions:
          - condition: trigger
            id:
              - GPSOn
...
        `

**Reply by bobthebuilder4711** [reply]
I've been monitoring the integration without adding anything. I just arrived with my Niro at 21:49. The location only updated around 10 minutes later at 21:57 and that's when all of the sensors where updated.

So either your automation has since been added to the integration or the car is sending more than the location now.

For me that means I'm going to try writing my own script for solar excess charging since I now know all the parameters. I'm already charging a stationary battery with solar excess with a self written script, so adding the car shouldn't be that hard to accomplish.

---

Last night I also noticed that all sensors for both cars switched to "unavailable" at around 22:00 and when I tried using the android app, it told me that I had hit the maximum amount of requests per day.

Seems like an update interval of 15 minutes and two cars was enough to hit the 200 early. Every 15 minutes is almost 100 in 24 hours. And with two cars it's almost 200. So looking at the app a few times and stopping charging for both cars seems to have been too much.

Will see if that changes once the Cee'd is going back to the leasing partner. For now I've set the interval to 20 minutes and that should give us enough room to get through the day.

**Reply by voravin** [reply]
I have a Kia Sportage PHEV, Is there any binary sensor that can inform when the car is powered on. The sensor ; binary_sensor.sportage_engine unfortunately only works if the car engine is powered on and not when the car is running in EV mode and powers off.

**Reply by bobthebuilder4711** [reply]
I'm guessing that there is no sensor like you want.

Is your AC ever off while the car is on? Because that sensor is there and could be used to check if the car is on or off.

**Reply by voravin** [reply]
There is the sensor for the AC but when running on EV mode, the AC is generally off unless it's really hot or in winter to defog the interior. There must be a sensor that monitors that the Kia has been powered off but maybe the integration has missed it.

**Reply by cdnninja** [reply]
Look at the raw data in the Data sensor and see if something fits for that. If so can create a new sensor for you with that data.

**Reply by bobthebuilder4711** [reply]
There is an "ign3" for my Niro that is currently set to false. Could that be the ignition? Does the Sportage have that field too?

**Reply by bobthebuilder4711** [reply]
But I still don't think that this would solve your problem. The car only sends an update to the servers when it is turned off. So to find out if the car is on at a given moment would mean that you'd have to make a forced update. And since you don't know when the car is turned on, you'd have to constantly poll it via forced updates, which isn't good for the battery if the car is off and might even get you temporary blocked from the service since you'd be polling way too often.

**Reply by voravin** [reply]
That's the problem.. I've found another way which I am testing.. Not idea but apparently (yet to be confirmed), when the car is powered off, it sends updated data to the servers so Kia connect has updated info. I've written an automation which polls the cache server with the odometer and check the reading against the last held and if it has changed assumes the car must have been powered on and then off as its now static. Currently I poll every 5min so data such as location is 5min out but I can live with that.. I'll then adjust the poll value to what I can get away with.. Will see if it works as intended later today.

**Reply by bobthebuilder4711** [reply]
How is that another way? Only when the car is powered off does it send an update to the servers. So polling the servers will only ever give you the last time the car was powered off. And that can be multiple hours ago even if you poll every 5 minutes. But you will never get an update while the car is on, which was your initial request. Could it be that you want to do something completely different if it is enough to know when the car has been turned off?

**Comment by busywait** [comment]
![Screenshot_20241219_000855_Home Assistant](https://github.com/user-attachments/assets/cfef0082-3649-49da-9144-888e98685f56)

**Reply by adorobis** [reply]
The kia app allows to refresh location data manually so it should be possible to get it updated by the API. Door lock status gets updated much more frequently so maybe we could force location update as well e.g. as a dedicated service in HA?

**Comment by Marcussjogren** [comment]
Old thread - but I thought I'd reuse it rather than starting a new one on the same matter.

Is there any reason to why the lower limit of updating cached values don't go below 15min? As far as I understand, this only queries Kias API and fetches those values, not getting them from the vehicle.

Such operation should be able to support to update at least once a minute, unless Kia is doing some sort of throttling?

**Reply by cdnninja** [reply]
The throttling exists.  You will max out api calls.  You can do one off calls via automation more frequently if desired.

**Reply by Marcussjogren** [reply]
> The throttling exists. You will max out api calls. You can do one off calls via automation more frequently if desired.

Do you have any idea of how many API-calls that are allowed to the cloud-data (not vehicle refresh)? I've read about 200 per day, but those posts were a couple of years old.

For me, I think 5min polls would be perfect. I've used it for my VW Passat and that gives a reasonable delay in changes.

**Reply by adorobis** [reply]
> The throttling exists. You will max out api calls. You can do one off calls via automation more frequently if desired.

But is there a way to request for location update? It will only update in the integration if I open the KIA app and get updated location. It was also confirmed by KIA support that that's the way it is. But it also means that there must be a possible API call to refresh the location so we could at least have the ability to refresh it from HA as well. Do do I understand it wrongly?

**Reply by bobthebuilder4711** [reply]
The integration has an action for that. Go to Developer tools->action and select "force update" from the list there. Select your car and then click "perform action".

If you want that on a dashboard then you need to create a button that performs that action.

**Reply by adorobis** [reply]
> The integration has an action for that. Go to Developer tools->action and select "force update" from the list there. Select your car and then click "perform action".
> 
> If you want that on a dashboard then you need to create a button that performs that action.

Oh, cool! How come I have not noticed that earlier. Many thanks!
